banyan/src/samples.py

```diff
@@ -1,17 +1,255 @@
+# Base.isnothing(s::Sample) = s.rate == -1
+
+# Caching samples with same statistics
+
+# A sample with memoized statistics for 
+# Must be mutable so that the Future finalizer runs
+class SampleForGrouping:
+    def __init__(self, future: Future, sample: Any, keys: List[Any], axes: List[int]):
+        self.future = future
+        self.sample = sample
+        self.keys = keys
+        self.axes = axes
+
+# Note that filtered_to's sample might be a vector
+
+# This functions is for retrieving a sample of a future with same
+# statistics properties with this key. Note that this function is not
+# type-stable and its return type isn't knowable so it _will_ result
+# in type inference. The best way to deal with that is to make sure to pass
+# the result of calling this (or even the other `sample` functions) into a
+# separate function to actually process its statistics. This creates a function
+# barrier and type inference will still happen at run time but it will only
+# happen once.
+def sample_for_grouping(f:Future, keys:List[Any], f_sample:Any)->SampleForGrouping:
+    # keys::Vector{K} = keys[1:min(8,end)]
+    # global same_statistics
+    # f_sample 
+    # # TODO: In the future we may need to allow grouping keys to be shared between futures
+    # # with different sample types but same grouping keys and same statistics.
+    # # For example, we might do an array resizing operation and keep some dimensions
+    # # that can still be having the same grouping keys. If we do this, we should immediately
+    # # ge ta type error here and will have to change the T to an Any.
+    # res::Dict{K,T} = Dict{K,T}()
+    # for key in keys
+    #     f_map_key::Tuple{ValueId,K} = (f.value_id, key)
+    #     s::T = haskey(same_statistics, f_map_key) ? same_statistics[f_map_key]::T : f_sample
+    #     res[key] = s
+    # end
+    return SampleForGrouping(f, f_sample, keys, sample_axes(f_sample))
+
+def sample_for_grouping(f:Future, keys:List[Any]):
+    return sample_for_grouping(f, keys, sample(f))
+
+def sample_for_grouping(f:Future, key:Any):
+    return sample_for_grouping(f, K[key])
+
+# TODO: Not sure what to do for below line
+# function sample_for_grouping(f::Future, ::Type{K}) where {K} sample_for_grouping(f, convert(Vector{K}, get_location(f).sample.groupingkeys)::Vector{K}) end
+
+def sample_for_grouping(f:Future):
+    return sample_for_grouping(f, int)
+
+class SampleComputationCache:
+    def __init__(computation: Dict[int, Any], same_keys: Dict[int, List[int]]):
+        self.computation = computation
+        self.same_keys = same_keys
+
+sample_computation_cache = SampleComputationCache({}, {})
+
+def get_sample_computation_cache()->SampleComputationCache
+    global sample_computation_cache
+    return sample_computation_cache
+
+def insert_in_sample_computation_cache(cache:SampleComputationCache, key:int, other_key:int)
+    if key not in cache.same_keys:
+        cache.same_keys[key] = UInt[key, other_key] # TODO: What is UInt?
```

Comment:
unsigned int

In Julia, this is the syntax to create a list of a particular type, so this is a list of unsigned ints. In python, lists are not typed, so we would just drop the `UInt`